src/pygeoprocessing/geoprocessing.py

```diff
@@ -535,6 +535,59 @@ def raster_calculator(
                 pass
 
 
+def raster_reduce(function, raster_path, initializer, band=1):
```

Comment:
I can see the parallels to python's reduce, which is neat!  A couple thoughts:

1. What would you think about using the `(path, band)` notation, for consistency with other raster operations in this file?
2. Would it be worth having the path/band tuple as the first parameter?  I see that most functions in `pygeoprocessing` that involve rasters do this, but also the function as it stands is more like the stdlib reduce.  What do you think?
3. One thing that's interesting and convenient about the `functools` reduce is that the `initializer` is optional.  Would that be worth including here as well?
4. Since this is a wrapper around `iterblocks`, I wonder if it would be helpful to also expose the `largest_block` parameter, in case the user knows something about their system and wants to fiddle with the block size.

Reply:
1. Sure!
2. I did put the arguments in this order to resemble `functools.reduce`, but consistency with pygeoprocessing is an equally valid reson. I don't really have a preference either way. 
4. Good idea!

Reply:
3: I think the `initializer` is necessary with the current design, but a maybe not with a different design.

With `functools.reduce(function, iterable, initializer)`, the initializer is optional because you can infer it. If `initializer` isn't provided, it's set to the return value of the first `function` call. 

But `raster_reduce` has two parts: it aggregates each block to a single value and then reduces _those_ values to a single value. In the current implementation, the `function` does both of those operations. For instance with the histogram example, the correct value of `initializer` to infer would be the result of `numpy.histogram` called on the first block. But there's no way to know that because `numpy.histogram` is wrapped up in the `function`.

But it would be possible if we split out the block aggregation and reducing operations into separate functions, something like:
```
def raster_reduce(raster_path, block_aggregate_fn, raster_reduce_fn, initializer=None):
    # reduce each block to a single value
    block_values = map(block_aggregate_fn, iterblocks(raster_path))
    if initializer is not None:
        block_values = [initializer] + block_values
    # reduce that series of block values to a single value
    return functools.reduce(raster_reduce_fn , block_values)
```
Then the histogram example could be written as
```
def block_histogram(block): return numpy.histogram(block, bins=10)[0]
def sum(x, y): return x + y
raster_reduce(raster_path, block_histogram, sum)
```
This is kind of hard to wrap my head around, does that make any sense?

Reply:
Aha!  Yes, you're absolutely right.  Thanks for walking me through this.  The return type of `function` might not be an array the size and shape (or even type) of `block`, so in general we can't just take the first block and use that as the initializer.

Ah, higher-order functions are so interesting!  I think we should stick with your original formulation with the initializer so that we minimize the number of functions one has to understand in order to get this working in code.  Conceptually the combination of map and reduce in the alternate implementation makes an interesting distinction, but I think the original implementation might be more straightforward.

Reply:
I had one other thought that I wonder what you think about:  would it make sense or be useful to optionally filter out nodata values?  We could always write our reduce function to handle nodata, but that requires knowing the nodata ahead of time when it could easily be read and the block indexed within `raster_reduce`.  If you agree, would it also make sense to flatten out the block for cases where we are not doing nodata filtering?  Just a thought!

Reply:
Good idea!